### DISCUSSION TITLE
Welcome to TMC2209_Raspberry_Pi Discussions!

### DISCUSSION THREAD

**Discussion by Chr157i4n** [Chr157i4n/PyTmcStepper#21] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Jim531980** [comment]
Dear Chr157i4n,

Congratulation for your work, I would like use it for one of my project. However, the step number for one revolution is like a constant in your code. It is set to 200 ( _fullsteps_per_rev = 200) and I would like to use a stepper motor with a different step number. Do you think it is possible that it could be parameter in the construtor of TMC2209?

Best regards.

**Reply by Chr157i4n** [reply]
Hello Jim and thank you. 

Is this the way you wanted it?
https://github.com/Chr157i4n/TMC2209_Raspberry_Pi/commit/69fe558447be4902bfe49826498d6fbfa2f4ab42

Otherwise you are welcome to make a pull request

**Reply by Jim531980** [reply]
Dear Chr157i4n,

Yes, this is the way I wanted but after thinking again on it, I think that other parameters should be configurable in the constructor.
I think that all parameters directly linked to stepper motor should be configurable in the constructor :
- Number of steps over one turn (covered by your previous message)
- Maximum current (linked to stepper motor coils) 
- Maximum speed (linked to stepper motor type)
- Maximum acceleration (linked to stepper motor type)
- ... (Other parameters I don't think about)

From my point of view, I think that I need to firstly create a class Motor to define the parameters of the stepper motor and then send these parameters to the constructor of the TMC2209. 

I am not an oriented object programmer expert so I don't know if your agree with this approach.

**Reply by Chr157i4n** [reply]
As Python supports named arguments i dont see such a big problem with having a lot of arguments in the constructor.
When you create an object of the class, you only have to use the arguments that you really need:
```python
tmc = TMC_2209(21, 16, 20, serialport="/dev/ttyAMA0", skip_uart_init=True)
```

One solution would be to use a settings data class (perhaps by using this https://docs.python.org/3/library/dataclasses.html) like you suggested.

Another solution would be to have proper setter/getter function for each value (which i would prefer):

- [@property decorator](https://docs.python.org/3/library/functions.html#property)
- [@getter decorator](https://docs.python.org/3/library/functions.html#property.getter)
- [@setter decorator](https://docs.python.org/3/library/functions.html#property.setter)

I am not sure which is the best solution. I am open to suggestions.

**Reply by Chr157i4n** [reply]
in the dev branch version the fullsteps_per_rev attribute can now be accessed as a property:

https://github.com/Chr157i4n/TMC2209_Raspberry_Pi/blob/22b1703f58d45fb7232782fc3c4ccac23bdc65b7/src/tmc_driver/tmc_220x.py#L79C4-L88C67
